File: JerezSur-Inmobiliaria/src/main/java/com/jerezsur/inmobiliaria/models/Usuario.java
package com.jerezsur.inmobiliaria.models;

import java.time.LocalDateTime;
import java.util.List;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.jerezsur.inmobiliaria.models.enums.AuthProvider;
import com.jerezsur.inmobiliaria.models.enums.OrigenUsuario;
import com.jerezsur.inmobiliaria.models.enums.Role;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Builder.Default;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Entity
@Table(name = "usuarios")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Usuario {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Email
    @Column(unique = true, nullable = true)
    private String email;

    @Column(unique = true, nullable = true) // Ahora opcional para permitir OAuth
    private String telefono;

    @NotBlank
    private String nombre;

    private String apellidos;

    @Column(nullable = true)
    private String password;

    @Column(unique = true, nullable = true)
    private String dni;

    private String imagenPerfilUrl;

    @Default
    private Boolean verified = false;

    private String comentarios;

    // --- ESTADO DE LA CUENTA ---
    @Default
    @Column(nullable = false)
    private Boolean cuentaActivada = false;

    @Enumerated(EnumType.STRING)
    @Default
    private OrigenUsuario origen = OrigenUsuario.AUTOREGISTRO;
    // AUTOREGISTRO, OAUTH, CRM_TRABAJADOR, WEB_CITA, WEB_VENTA

    // --- SEGURIDAD Y ROLES ---
    @Enumerated(EnumType.STRING)
    @Default
    private Role role = Role.ROLE_NOROL;

    @Default
    private Boolean cambiarPasswd = true;

    private AuthProvider provider;
    private String providerId;

    // --- RELACIONES DE PERFIL ---
    @OneToOne(mappedBy = "usuario", cascade = CascadeType.REMOVE, orphanRemoval = true)
    @ToString.Exclude
    @JsonIgnore
    private Trabajador trabajador;

    @OneToOne(mappedBy = "usuario", cascade = CascadeType.REMOVE, orphanRemoval = true)
    @JsonIgnore
    private Interesado interesado;

    @OneToOne(mappedBy = "usuario", cascade = CascadeType.REMOVE, orphanRemoval = true)
    @JsonIgnore
    private Vendedor vendedor;

    @OneToMany(mappedBy = "usuario")
    @ToString.Exclude
    @JsonIgnore
    private List<Cita> citas;

    private LocalDateTime fechaEliminacion;

    @CreationTimestamp
    private LocalDateTime fechaRegistro;

    @UpdateTimestamp
    private LocalDateTime fechaUltimaActualizacion;
}

File: JerezSur-admin/src/types/usuario.ts
// ==========================================
// ENUMS Y TIPOS AUXILIARES
// ==========================================

export type OrigenUsuario =
  | 'AUTOREGISTRO'
  | 'OAUTH'
  | 'CRM_TRABAJADOR'
  | 'WEB_CITA'
  | 'WEB_VENTA';

export type Role =
  | 'ROLE_ADMIN' // Administradores de la Web (YO)
  | 'ROLE_TRABAJADOR' // Trabajadores de JerezSur
  | 'ROLE_INTERESADO' // Interesados y/o compradores de inmuebles
  | 'ROLE_VENDEDOR' // Vendedores de inmuebles
  | 'ROLE_AMBOS' // Usuarios que son tanto interesados como vendedores
  | 'ROLE_NOROLE'; // Usuarios sin rol asignado (posible estado inicial o error)

export type AuthProvider = 'LOCAL' | 'GOOGLE' | 'FACEBOOK' | 'APPLE'; // Ajusta según tu backend

// ==========================================
// INTERFAZ PRINCIPAL (EQUIVALENTE A LA ENTIDAD)
// ==========================================

export interface Usuario {
  id: number;
  nombre: string;
  cuentaActivada: boolean;
  verified: boolean;
  cambiarPasswd: boolean;
  origen: OrigenUsuario;
  role: Role;
  fechaRegistro: string;           // ISO 8601: "2026-05-27T08:42:00"
  fechaUltimaActualizacion: string; // ISO 8601

  // Campos opcionales (nullable = true o relaciones)
  email?: string;
  telefono?: string;
  apellidos?: string;
  password?: string; // Normalmente no se envía al frontend, pero se incluye por la entidad
  dni?: string;
  imagenPerfilUrl?: string;
  comentarios?: string;
  provider?: AuthProvider;
  providerId?: string;

  // IDs de relaciones (en el frontend es mejor manejar IDs que el objeto completo anidado)
  trabajadorId?: number;
  interesadoId?: number;
  vendedorId?: number;
  citasIds?: number[];
}

// ==========================================
// DTO PARA CREACIÓN / REGISTRO
// ==========================================

export interface NuevoUsuario {
  nombre: string;
  email?: string;
  telefono?: string;
  apellidos?: string;
  password?: string;
  dni?: string;
  origen?: OrigenUsuario;
}

// ==========================================
// SUB-PERFILES ESPECÍFICOS
// ==========================================

export interface TrabajadorPerfil {
  id: number;
  dni: string;
  cargo?: string;
  fechaInicioContrato: string; // ISO Date "YYYY-MM-DD"
  fechaFinContrato?: string;
  activo: boolean;
  observacionesLaborales?: string;
}

export interface InteresadoPerfil {
  id: number;
  estado: 'INTERESADO' | 'ACTIVO' | 'NEGOCIACION' | 'CONTRATADO' | 'DESCARTADO'; // Ajusta según tu EstadoComprador
  requiereHipoteca: boolean;
  detallesHipoteca?: string;
  presupuestoMaximo?: number;
  zonaInteres?: string;
  habitacionesMinimas?: number;
  banosMinimos?: number;
  tipoInmueblePreferido?: string;
  tipoBusqueda?: 'VENTA' | 'ALQUILER' | 'CUALQUIERA'; // Ajusta según tu TipoOperacion
  observaciones?: string;
}

export interface VendedorPerfil {
  id: number;
  observaciones?: string;
}

// ==========================================
// INTERFAZ EXTENDIDA PARA DETALLE
// ==========================================

export interface UsuarioDetalle extends Usuario {
  // Ahora el backend nos inyectará el objeto completo cuando lo pidamos por ID
  trabajador?: TrabajadorPerfil;
  interesado?: InteresadoPerfil;
  vendedor?: VendedorPerfil;
}